[addons][viz] cleanup API, improve "C" interface, not use void* direct as value

AlwinEsch opened the pull request:
## Description

This thought to make **Visualization**  in way equal to PVR API, so to have the "C" in separate header and helper class.

## User related:

Not user relevant

Only for all binary addon system requests may be for the user:
- Binary addon system revised to ensure higher security in its data exchange between addon and Kodi
- Revised binary addon system documentation for outside developers

Commit 1:
-----------------------------------
**[addons][viz] don't use Action with `void*` in "C" interface**

Before was there several parts where given via a Action call to addon and then changed to the separate C++ functions.

This now also change the "C" parts to have about every action a own function.

Commit 2:
-----------------------------------
**[addons][viz] don't use C++ class in C**

Before was the C++ class pointer exchanged between Addon and Kodi,
it was not used in Kodi but it breaks the ABI and can't compiled
in "C" only.

Reasons about:
- As the data value before was with void* where declare struct, int, bool... and was not safe against failures (if used in "C" only)
- Looks more clean now to see what is called, as before was only one function and easily a part can be overseen
- Better debugging

Commit 3:
-----------------------------------
**[addons][viz] add class VisualizationTrack to give VIS_TRACK in C++**

This is a translation class to give the "C" styled data in a "C++" class.
Thought to have it more clean and also copyable on addon.

By this is more safe in use on addon.

Commit 4:
-----------------------------------
**[addons][viz] separate "C" to own file**

This to hold "C++" and "C" independent and allow for other languages.

Also is a clang cleanup on include/kodi/addon-instance/Visualization.h done

Commit 5:
-----------------------------------
**[addons][viz] use ADDON_HARDWARE_CONTEXT and kodi::HardwareContext for `void*`**

There becomes on HardwareContext related calls (e.g. to get `ID3D11DeviceContext*` on Windows)
within visualization the ADDON_HARDWARE_CONTEXT about "C" and the kodi::HardwareContext about "C++"
used, instead of void* before.

The value itself still void*, only named to related type. This done to prevent inside
Kodi's API, the use of void* on addon as much as possible.

All places where the addon must use them should be with a name to have more clean.
This then also better if this parts used for other languages.

Commit 6:
-----------------------------------
**[addons][viz] cleanup visualization addon instance documentation**

To make it a bit cleaner, fix few doxygen parts and add few new text parts.

Commit 7:
-----------------------------------
**[addons][viz] increase visualization version to 3.0.0**


<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[addons][viz] don't use Action with void* in "C" interface

Before was there several parts where given via a Action call to addon and then changed to the separate C++ functions.

This now also change the "C" parts to have about every action a own function.

Reasons about:
- As the data value before was with void* where declare struct, int, bool... and was not safe against failures (if used in "C" only)
- Looks more clean now to see what is called, as before was only one function and easily a part can be overseen
- Better debugging

AlwinEsch committed:
[addons][viz] don't use C++ class in C

Before was the C++ class pointer exchanged between Addon and Kodi,
it was not used in Kodi but it breaks the ABI and can't compiled
in "C" only.

AlwinEsch committed:
[addons][viz] use ADDON_HARDWARE_CONTEXT and kodi::HardwareContext for void*

There becomes on HardwareContext related calls (e.g. to get ID3D11DeviceContext* on Windows)
within visualization the ADDON_HARDWARE_CONTEXT about "C" and the kodi::HardwareContext about "C++"
used, instead of void* before.

The value itself still void*, only named to related type. This done to prevent inside
Kodi's API, the use of void* on addon as much as possible.

All places where the addon must use them should be with a name to have more clean.
This then also better if this parts used for other languages.

AlwinEsch committed:
[addons][viz] cleanup visualization addon instance documentation

To make it a bit cleaner, fix few doxygen parts and add few new text parts.

AlwinEsch committed:
[addons][viz] increase visualization version to 3.0.0

AlwinEsch committed:
[addons][viz] separate "C" to own file

This to hold "C++" and "C" independent and allow for other languages.

Also is a clang cleanup on include/kodi/addon-instance/Visualization.h done

garbear commented on xbmc/guilib/GUIVisualisationControl.cpp:
```diff
@@ -298,21 +298,22 @@ void CGUIVisualisationControl::UpdateTrack()
-  track.rating      = tag->GetUserrating();
-
-  m_instance->OnAction(VIS_ACTION_UPDATE_TRACK, &track);
+  VIS_TRACK* track = new VIS_TRACK(); // allocate by his size
```
Can you be more descriptive about why this is allocated?

AlwinEsch replied:
Ah yeah, is about this place https://github.com/xbmc/xbmc/pull/18251/files#diff-3db4e9fc8c806bfb2460fe6b33d6b649R34

Maybe?:
- `Its size can exceed the available stack size, an explicit "new" is necessary.`

garbear replied:
And we can be a bit more specific about why it exceeds:

* `Its size can exceed the available stack size due to the 32K allocated for lyrics, so an explicit "new" is necessary.`

garbear replied:
@AlwinEsch after seeing the issue solved by https://github.com/xbmc/xbmc/pull/18264 I wonder if we shouldn't make the lyric size dynamic instead of moving the struct to the heap

AlwinEsch replied:
It's a bit the thing about safe "C" interface, but to have them allocated in matching size of lyrics text would be the best.

But not know how easily it can done, if they come only from Kodi and nowhere copied with the C++ part about on addon side. Then it can be a allocated pointer about.

Only becomes addon allowed to copy itself this class, where is done by 'CStructHdl` (https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h#L217-L282) and where makes a complete copy of "C" part for other place.

If for every value it becomes done by normal pointer instead of fixed array, must the code of related class do byself where can be much code text.

Can look to make one special alloc about them and the rest with "CStructHdl", hope that works.

garbear commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -29,6 +29,34 @@ namespace gui
+/// values.
+///
+/// This is basically a simple pointer which has to be changed to the desired
+/// format at the corresponding places using <b>`static_cast<...>(...)`</b>.
```
Helpful and explanatory documentation

garbear commented:
Looks good

AlwinEsch committed:
[addons][viz] add class VisualizationTrack to give VIS_TRACK in C++

This is a translation class to give the "C" styled data in a "C++" class.
Thought to have it more clean and also copyable on addon.

By this is more safe in use on addon.

AlwinEsch commented:
Have it updated and the data becomes now transfered in old way (which with less memory usage), on the C++ part by addon side becomes the "C" part copied to "C++" place. By this only the needed mem becomes in background created for.
